modules/swagger-codegen/src/main/resources/rust/api.mustache

```diff
@@ -39,26 +40,84 @@ impl<C: hyper::client::Connect>{{classname}} for {{classname}}Client<C> {
+                let val = match apikey.prefix {
+                    Some(ref prefix) => format!("{} {}", prefix, key),
+                    None => key,
+                };
+            {{#isKeyInHeader}}
+                auth_headers.insert("{{keyParamName}}".to_owned(), val);
+            {{/isKeyInHeader}}
+            {{#isKeyInQuery}}
+                auth_query.insert("{{keyParamName}}".to_owned(), val);
+            {{/isKeyInQuery}}
+            }
+        };
+{{/isApiKey}}
+{{#isBasic}}
+        {
+            match configuration.basic_auth {
+                Some(ref auth_conf) => {
+                    let auth = hyper::header::Authorization(
+                        hyper::header::Basic {
+                            username: auth_conf.0.to_owned(),
+                            password: auth_conf.1.to_owned(),
+                        }
+                    );
+                    auth_headers.insert("Authorization".to_owned(), format!("{}", auth));
+                }
+                None => {}
+            }
+        };
+{{/isBasic}}
+{{#isOAuth}}
+        {
+            match configuration.oauth_access_token {
+                Some(ref token) => {
+                    let auth = hyper::header::Authorization(
+                        hyper::header::Bearer {
+                            token: token.to_owned(),
+                        }
+                    );
+                    auth_headers.insert("Authorization".to_owned(), format!("{}", auth));
+                }
+                None => {}
+            }
+        };
+{{/isOAuth}}
+{{/authMethods}}
+{{/hasAuthMethods}}
         let method = hyper::Method::{{httpMethod}};
 
-        {{^hasQueryParams}}
-        let uri_str = format!("{}{{{path}}}", configuration.base_path{{#pathParams}}, {{baseName}}={{paramName}}{{#isListContainer}}.join(",").as_ref(){{/isListContainer}}{{/pathParams}});
-        {{/hasQueryParams}}
-        {{#hasQueryParams}}
-        let query = ::url::form_urlencoded::Serializer::new(String::new())
-            {{#queryParams}}
-            .append_pair("{{baseName}}", &{{paramName}}{{#isListContainer}}.join(","){{/isListContainer}}.to_string())
-            {{/queryParams}}
-            .finish();
-        let uri_str = format!("{}{{{path}}}{}", configuration.base_path, query{{#pathParams}}, {{baseName}}={{paramName}}{{#isListContainer}}.join(",").as_ref(){{/isListContainer}}{{/pathParams}});
-        {{/hasQueryParams}}
-
-        let uri = uri_str.parse();
+        let query_string = {
+            let mut query = ::url::form_urlencoded::Serializer::new(String::new());
+{{#queryParams}}
+            query.append_pair("{{baseName}}", &{{paramName}}{{#isListContainer}}.join(","){{/isListContainer}}.to_string());
+{{/queryParams}}
+{{#hasAuthMethods}}
+            for (key, val) in &auth_query {
+                query.append_pair(key, val);
+            }
+{{/hasAuthMethods}} 
+            format!("?{}", query.finish())
```

Comment:
Since the `?` will end up even for an empty query string, I would just put in the `format!` below and just return `query.finish()` here.